src/aria/private/grid/grid.ts

```diff
@@ -304,27 +344,78 @@ export class GridPattern {
   resetStateEffect(): void {
     const hasReset = this.gridBehavior.resetState();
 
-    // If the active state has been reset right after a focusout event, then
-    // we know it's caused by a row/cell deletion.
-    if (hasReset && this._maybeDeletion()) {
-      this._deletion.set(true);
+    if (hasReset) {
+      // If the active state has been reset right after a focusout event, then
+      // we know it's caused by a row/cell deletion.
+      if (this._maybeDeletion()) {
+        this._deletion.set(true);
+      } else {
+        this._stateStale.set(true);
+      }
     }
     // Reset maybe deletion state.
     this._maybeDeletion.set(false);
   }
 
-  /** Focuses on the active cell element. */
-  focusEffect(): void {
-    const activeCell = this.activeCell();
-    const hasFocus = this.isFocused();
+  /** Resets the focus to the active cell element or grid element. */
+  resetFocusEffect(): void {
+    const stateStale = this._stateStale();
+    if (!stateStale) return;
+
+    const isFocused = untracked(() => this.isFocused());
+    const isRoving = untracked(() => this.inputs.focusMode() === 'roving');
+    const activeCell = untracked(() => this.activeCell());
+
+    if (isRoving && activeCell !== undefined && isFocused) {
+      if (!activeCell.isFocused()) {
+        activeCell.focus();
+      }
+    }
+
+    this._stateStale.set(false);
+  }
+
+  /** Restore focus when a deletion happened. */
+  restoreFocusEffect(): void {
     const deletion = this._deletion();
-    const isRoving = this.inputs.focusMode() === 'roving';
-    if (activeCell !== undefined && isRoving && (hasFocus || deletion)) {
-      activeCell.element().focus();
+    if (!deletion) return;
 
-      if (deletion) {
-        this._deletion.set(false);
+    const isRoving = untracked(() => this.inputs.focusMode() === 'roving');
+    const activeCell = untracked(() => this.activeCell());
+
+    if (isRoving && activeCell !== undefined) {
+      if (!activeCell.isFocused()) {
+        activeCell.focus();
       }
     }
+
+    this._deletion.set(false);
+  }
+
+  /** Sets focus when active cell changed. */
+  focusEffect(): void {
+    const activeCell = this.activeCell();
```

Comment:
Is this effect meant to be called every time the active cell changes? IIRC, there was a bug in the datepicker example because changing the month would change the active item and cause focus to jump back into the grid